File: src/main/java/x/withlithum/neoware/game/player/PlayerDataUtil.java
package x.withlithum.neoware.game.player;

import lombok.extern.slf4j.Slf4j;
import net.minestom.server.entity.Player;
import org.jspecify.annotations.NullMarked;
import x.withlithum.neoware.data.SavedPosition;
import x.withlithum.neoware.data.player.PlayerInfo;
import x.withlithum.neoware.data.player.PlayerStatus;
import x.withlithum.neoware.game.inventory.InventoryResolver;
import x.withlithum.neoware.server.commands.PermissionRank;

@Slf4j
@NullMarked
public final class PlayerDataUtil {
    public static final int VERSION = 2;

    public static void recoverPlayer(Player player, PlayerInfo info) {
        player.setPermissionLevel(info.rank().ordinal());
        info.status().apply(player);
        player.teleport(info.position().asPos());
        player.setRespawnPoint(info.respawnPoint().asPos());
        player.getInventory().clear();

        if (info.items() != null) {
            if (player.getInventory().getSize() != info.items().size()) {
                log.warn("Inventory size mismatch for player '{}' ({}): stored {} versus {}",
                    player.getUsername(),
                    player.getUuid(),
                    info.items().size(),
                    player.getInventory().getSize());
                log.warn("Not recovering inventory");
            } else {
                InventoryResolver.recover(info.items(), player.getInventory());
            }
        }
    }

    public static PlayerInfo createPlayerInfo(Player player) {
        var items = InventoryResolver.resolve(player.getInventory());

        return new PlayerInfo(VERSION,
            PermissionRank.values()[player.getPermissionLevel()],
            PlayerStatus.create(player),
            SavedPosition.create(player.getPosition()),
            SavedPosition.create(player.getRespawnPoint()),
            items
        );
    }
}

File: adventure/src/main/kotlin/x/withlithum/neoware/adventure/server/storage/PlayerDataUtil.kt
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.adventure.server.storage

import io.github.oshai.kotlinlogging.KotlinLogging
import net.minestom.server.entity.Player
import net.minestom.server.inventory.AbstractInventory
import net.minestom.server.item.ItemStack
import x.withlithum.neoware.adventure.content.item.AdventureItemManager
import x.withlithum.neoware.data.game.ItemRef
import x.withlithum.neoware.data.player.PlayerInfo
import x.withlithum.neoware.data.player.PlayerStatus
import x.withlithum.neoware.server.commands.PermissionRank
import x.withlithum.neoware.util.results.NeoResult

object PlayerDataUtil {
    const val VERSION = 3

    private val logger = KotlinLogging.logger {}

    fun recoverInventory(refs: List<ItemRef>,
                         inventory: AbstractInventory,
                         itemManager: AdventureItemManager) {
        inventory.clear()

        for (i in refs.indices) {
            val ref = refs[i]
            when (val result = itemManager.resolveRef(ref)) {
                is NeoResult.Error -> result.logWarn(logger)
                is NeoResult.Ok -> inventory.setItemStack(i, result.value, false)
            }
            inventory.update()
        }
    }

    fun recoverPlayer(player: Player, playerInfo: PlayerInfo, itemManager: AdventureItemManager) {
        player.permissionLevel = playerInfo.rank.ordinal
        playerInfo.status.apply(player)

        if (playerInfo.items != null) {
            val items = playerInfo.items!!
            if (player.inventory.size != items.size) {
                logger.warn {"Inventory size mismatch (player ${player.inventory.size} vs data ${items.size})" }
            } else {
                recoverInventory(items, player.inventory, itemManager)
            }
        }
    }

    fun storeInventory(inventory: AbstractInventory, itemManager: AdventureItemManager): List<ItemRef> {
        val contents = Array(inventory.size) { ItemStack.AIR }
        inventory.copyContents(contents)
        val refs = ArrayList<ItemRef>(contents.size)

        for (i in contents.indices) {
            val item = contents[i]
            when (val result = itemManager.createRef(item)) {
                is NeoResult.Error -> {
                    result.logWarn(logger)
                    refs.add(ItemRef.AIR)
                }
                is NeoResult.Ok -> refs.add(result.value)
            }
        }

        return refs
    }

    fun storePlayer(player: Player, itemManager: AdventureItemManager): PlayerInfo {
        val inventory = storeInventory(player.inventory, itemManager)

        val permission = if (player.permissionLevel > 4) {
            PermissionRank.WHEEL
        } else {
            PermissionRank.entries[player.permissionLevel]
        }

        return PlayerInfo(VERSION,
            permission,
            PlayerStatus.create(player),
            inventory)
    }
}